File: config/src/main/java/com/typesafe/config/impl/AbstractConfigNode.java
/**
 *   Copyright (C) 2015 Typesafe Inc. <http://typesafe.com>
 */
package com.typesafe.config.impl;

import java.util.Collection;

import com.typesafe.config.ConfigOrigin;
import com.typesafe.config.parser.ConfigNode;

abstract class AbstractConfigNode implements ConfigNode {
    abstract Collection<Token> tokens();

    final public String render() {
        StringBuilder origText = new StringBuilder();
        Iterable<Token> tokens = tokens();
        for (Token t : tokens) {
            origText.append(t.tokenText());
        }
        return origText.toString();
    }

    @Override
    final public boolean equals(Object other) {
        return other instanceof AbstractConfigNode && render().equals(((AbstractConfigNode) other).render());
    }

    @Override
    final public int hashCode() {
        return render().hashCode();
    }

    @Override
    public ConfigOrigin origin() {
        Collection<Token> tok = tokens();
        if (tok.isEmpty())
            return null;
        else
            return tok.iterator().next().origin();
    }
}

File: config/src/main/scala/com/typesafe/config/impl/AbstractConfigNode.scala
/**
 *   Copyright (C) 2015 Typesafe Inc. <http://typesafe.com>
 */
package com.typesafe.config.impl

import com.typesafe.config.parser.ConfigNode
import java.{ util => ju }

abstract class AbstractConfigNode extends ConfigNode {

    protected[impl] def tokens(): ju.Collection[Token]

    override final def render: String = {
        val origText = new StringBuilder
        import scala.collection.JavaConverters._
        for (t <- tokens().asScala) {
            origText.append(t.tokenText)
        }
        origText.toString
    }

    override final def equals(other: Any): Boolean =
        other.isInstanceOf[AbstractConfigNode] && render == other
            .asInstanceOf[AbstractConfigNode]
            .render

    override final def hashCode: Int = render.hashCode
}
